core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/WorkflowWorker.scala

```diff
@@ -0,0 +1,289 @@
+      workerStateManager.transitTo(Paused)
+      reportState()
+    case UpdateInputLinking(identifier, inputNum) =>
+      logger.info(s"received register input for ${this.identifier}")
+      sender ! Ack
+      tupleProducer.registerInput(identifier, inputNum)
+    case LocalBreakpointTriggered() =>
+      dataProcessor.breakpoints.foreach { brk =>
+        if (brk.isTriggered)
+          dataProcessor.unhandledFaultedTuples(brk.triggeredTupleId) =
+            new FaultedTuple(brk.triggeredTuple, brk.triggeredTupleId, brk.isInput)
+      }
+      context.parent ! ReportState(WorkerState.LocalBreakpointTriggered)
+    case other =>
+  }
+
+  def getOldWorkerState: WorkerState.Value = {
+    workerStateManager.getCurrentState match {
+      case UnInitialized =>
+        WorkerState.Uninitialized
+      case Ready =>
+        WorkerState.Ready
+      case Running =>
+        WorkerState.Running
+      case Pausing =>
+        WorkerState.Pausing
+      case Paused =>
+        WorkerState.Paused
+      case Completed =>
+        WorkerState.Completed
+      case Recovering =>
+        WorkerState.Running
+    }
+  }
+
+  def reportState(): Unit = context.parent ! ReportState(getOldWorkerState)
+
+  def oldControlMessageHandler: Receive = {
+    case Start =>
+      sender ! Ack
+      if (workerStateManager.getCurrentState != Ready) {
+        logger.info(
+          s"unexpected Start message when worker is in ${workerStateManager.getCurrentState}"
+        )
+      } else if (operator.isInstanceOf[ISourceOperatorExecutor]) {
+        dataProcessor.appendElement(EndMarker())
+        dataProcessor.appendElement(EndOfAllMarker())
+        workerStateManager.transitTo(Running)
+        reportState()
+      } else {
+        logger.info("unexpected Start message for non-source operator!")
+      }
+    case Pause =>
+      promiseManager.execute(PromiseInvocation(null, WorkerPause()))
+      workerStateManager.transitTo(Pausing)
+      reportState()
+    case Resume =>
+      pauseManager.resume()
+      workerStateManager.transitTo(Running)
+      reportState()
+    case AckedWorkerInitialization(recoveryInformation) =>
+      operator.open()
+      workerStateManager.transitTo(Ready)
+      reportState()
+    case QueryState =>
+      reportState()
+    case QueryStatistics =>
+      sender ! ReportStatistics(
+        WorkerStatistics(getOldWorkerState, getInputRowCount(), getOutputRowCount())
+      )
+    case QueryTriggeredBreakpoints =>
+      val toReport = dataProcessor.breakpoints.filter(_.isTriggered)
+      if (toReport.nonEmpty) {
+        toReport.foreach(_.isReported = true)
+        sender ! ReportedTriggeredBreakpoints(toReport)
+      } else {
+        throw WorkflowRuntimeException(
+          WorkflowRuntimeError(
+            "no triggered local breakpoints but worker in triggered breakpoint state",
+            "WorkerBase:allowQueryTriggeredBreakpoints",
+            Map()
+          )
+        )
+      }
```

Comment:
should this be allowed in unitialized state? will dataprocessor.breakpoint not throw null pointer

Reply:
dataprocessor.breakpoints will be an empty array at the beginning, null pointer exception won't be thrown.